lib/csv.ts

```diff
@@ -0,0 +1,136 @@
+  var csvContent = csv.headerRow + csv.body
+  if (!csvContent.match(/^data:text\/csv/i)) {
+    csvContent = 'data:text/csv;charset=utf-8,' + csvContent // use 'data:text/csv;charset=utf-8,\ufeff', if you consider using the excel
+  }
+  var data = encodeURI(csvContent)
+
+  var link = document.createElement('a')
+  link.href = data
+  link.download = csv.filename
+
+  document.body.appendChild(link)
+  link.click()
+  document.body.removeChild(link)
+}
+
+export const processFileForWorkspaceImport = (
+  file: File,
+): Promise<string[]> => {
+  return new Promise((resolve, reject) => {
+    const fileType = file.type
+    const fileName = file.name.toLowerCase()
+
+    if (fileType === 'application/json' || fileName.endsWith('.json')) {
+      const reader = new FileReader()
+      reader.onload = () => {
+        try {
+          const jsonData = JSON.parse(reader.result as string)
+          const values = extractFromJSON(jsonData)
+          if (values.length === 0) {
+            reject(new Error(`No 'did' or 'uri' found in ${file.name}`))
+          }
+          resolve(values)
+        } catch (error) {
+          reject(new Error(`Invalid JSON file: ${file.name}`))
+        }
+      }
+      reader.onerror = () => reject(new Error(`Failed to read ${file.name}`))
+      reader.readAsText(file)
+    } else if (fileType === 'text/csv' || fileName.endsWith('.csv')) {
+      const reader = new FileReader()
+      reader.onload = () => {
+        try {
+          const csvData = reader.result as string
+          const values = extractFromCSV(csvData)
+          if (values.length === 0) {
+            reject(new Error(`No 'did' or 'uri' found in ${file.name}`))
+          }
+          resolve(values)
+        } catch (error) {
+          reject(new Error(`Invalid CSV file: ${file.name}`))
+        }
+      }
+      reader.onerror = () => reject(new Error(`Failed to read ${file.name}`))
+      reader.readAsText(file)
+    } else {
+      reject(new Error(`Unsupported file type: ${file.name}`))
+    }
+  })
+}
+
+export const extractFromCSV = (data: string): string[] => {
+  const rows = data.split('\n').map((row) => row.trim())
+  const [header, ...content] = rows
+
+  if (!header) return []
+
+  // In case header names are quoted, we want to exclude those quotes before check
+  const headers = header.split(',').map((col) => {
+    const trimmed = col.trim()
+    return trimmed.startsWith('"') && trimmed.endsWith('"')
+      ? trimmed.slice(1, -1)
+      : trimmed
+  })
+  const didIndex = headers.indexOf('did')
+  const uriIndex = headers.indexOf('uri')
+
+  if (didIndex === -1 && uriIndex === -1) return []
+
+  return content
+    .map((row) => {
+      const columns = row.split(',').map((col) => {
+        const trimmed = col.trim()
+        return trimmed.startsWith('"') && trimmed.endsWith('"')
+          ? trimmed.slice(1, -1)
+          : trimmed
+      })
+      return columns[didIndex] || columns[uriIndex]
```

Comment:
Looks like the row parsing and header parsing contain a lot of duplication

Reply:
cleaned em up thanks!